cannon-kona: Add support for building kona prestate and kona-host

ajsutton opened the pull request:
**Description**

Adds a just file to build kona prestates and Kona-host using a particular Kona version. While the build can take 5-10 minutes we can leverage CircleCI caching to avoid building it when we haven't changed anything about the build process. Additionally it's designed so that the hash is provided to configure games but the prestate itself doesn't have to be available unless tests actually use it.

There's a skipped test for cannon-kona pre-interop which is failing because sys go provides ws:// URLs for the L2 EL and Kona-host only support http:// URLs.  We'll need to follow up to either fix kona or change sys go.


**Metadata**

Part of https://github.com/ethereum-optimism/optimism/issues/17764

ajsutton committed:
kona: Add a justfile to build a specific version of kona

ajsutton committed:
Try building on a machine

ajsutton committed:
Cache properly

ajsutton committed:
Remove REPO_ROOT

ajsutton committed:
Build pre-interop and interop kona prestates

ajsutton committed:
Set the cannon-kona prestate in interop tests.

Inphi commented on kona/justfile:
```diff
@@ -0,0 +1,48 @@
+    if [[ -d kona ]]; then
+      cd kona
+      git checkout -f "{{KONA_VERSION}}"
+      git reset --hard HEAD
```
checkout with -f already removes unwanted files.
```suggestion
```

ajsutton replied:
The docs suggest it only throws away local changes that are in the way of the update, not that it throws away all changes.

```
-f, --force
           When switching branches, proceed even if the index or the working tree differs from HEAD, and even if there are untracked files in the way. This is used to throw away local
           changes and any untracked files or directories that are in the way.

           When checking out paths from the index, do not fail upon unmerged entries; instead, unmerged entries are ignored.
```

Inphi replied:
I see. It's good keep the reset then to remove any generated files.

ajsutton committed:
ci: Build kona-host

ajsutton committed:
Test kona-host can execute

ajsutton committed:
Fix rust image version

ajsutton committed:
Use two separate machines

ajsutton committed:
No remote docker for machine executor

ajsutton committed:
Update rust

ajsutton committed:
Install clang

ajsutton committed:
Fix machine config

ajsutton committed:
Use default docker image since mise is installing rust

ajsutton committed:
Set kona-host and prestate URL in challenger config

ajsutton committed:
Remove test-kona step

ajsutton committed:
Try not installing clang

ajsutton committed:
Change justfile

ajsutton committed:
Really force a change

ajsutton committed:
Back to installing clang

ajsutton committed:
Restore cannon prestate for kurtosis tests

ajsutton committed:
Remove forced change var

ajsutton committed:
Ensure tests that don't use kona can run even if kona components haven't been built.

Add skipped test to actually run cannon-kona game. Doesn't pass yet because kona-host doesn't support ws:// URLs which are used by sysgo.

ajsutton committed:
Add cannon-kona config for interop challenger.

ajsutton committed:
Only add cannon-kona games when feature toggle is enabled.

Inphi commented on .circleci/config.yml:
```diff
@@ -1929,6 +1929,55 @@ jobs:
+      - checkout-from-workspace
+      - restore_cache:
+          name: Restore kona cache
+          key: kona-prestate-{{ checksum "./kona/justfile" }}
```
The cached prestate won't be bust by changes to the kona source code. For example, if we update the kona tag to `kona-client/v1.3.0` and the justfile hasn't changed, then we continue to use the cached prestate for v1.2.0.
How about using the kona version we're building as the cache key? Since it's a reproducibile build it should suffice.

ajsutton replied:
Yeah that's why I made sure the kona version we use is inside the justfile.

Inphi replied:
I see. I didn't realize this is not the justfile in the kona repo. Looks good.

Inphi commented on .circleci/config.yml:
```diff
@@ -1929,6 +1929,55 @@ jobs:
+          key: kona-prestate-{{ checksum "./kona/justfile" }}
+      - run:
+          name: Build kona prestates
+          command: just build-prestates
```
If we use the kona version as the cache key, then it should be explicitly set here when invoking the just target.

Inphi commented on .circleci/config.yml:
```diff
@@ -1929,6 +1929,55 @@ jobs:
+      - checkout-from-workspace
+      - restore_cache:
+          name: Restore kona host cache
+          key: kona-host-{{ checksum "./kona/justfile" }}
```
same thing here. Worth revising how the kona host is cached.

Inphi commented on kona/justfile:
```diff
@@ -0,0 +1,80 @@
+KONA_DIR := `realpath .`
+KONA_VERSION := "kona-client/v1.2.0"
+KONA_PRESTATE_HASH := "0x03723aeb1fe69eebd6b0155a4570f5944e9a2c1d918183a014b6da2fc2d1b5e5"
+KONA_INTEROP_PRESTATE_HASH := "0x03dd682a7c55084b42c4f4711f2c501126ed8dc05808c7ef19537dbf57bcbed7"
```
rather than hardcode this, we could read the hash from the generated prestate-artifacts-cannon/prestate-proof.json. It'll make it easier to bump the KONA_VERSION in the future.

Inphi commented on kona/run-kona-host.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+# This is a shim for kona-host used as the --cannon-kona-server value when running op-challenger in acceptance tests
```
We could be more selective in enabling when `presets.WithCannonKona()` is used. Only tests that actually  run kona in the VM should have that specified.

If the op-challenger only configured the cannon-kona trace type when an acceptance test enables `presets.WithCannonKona()`, would that eliminate the need for this shim?

ajsutton replied:
We only actually enable cannon-kona games on specific tests, but the config is added to the challenger config. I'll take a look at whether we can only add that conditionally as well, but will depend on whether we can get access to the required information at that point.

ajsutton replied:
Managed to only set the cannon kona config when the game is actually being added so we don't need the run-kona-host.sh script now.

Inphi commented on kona/justfile:
```diff
@@ -0,0 +1,80 @@
+KONA_DIR := `realpath .`
+KONA_VERSION := "kona-client/v1.2.0"
```
v1.2.0 no longer exists. You can use v1.2.2 instead.

Inphi commented on op-devstack/sysgo/superroot.go:
```diff
@@ -225,10 +228,38 @@ func getSuperRoot(t devtest.CommonT, o *Orchestrator, timestamp uint64, supervis
+}
+
+func findKonaVariable(t devtest.CommonT, name string) string {
+	konaJustfilePath := "kona/justfile"
```
nit: Rather than parsing the justfile, the kona variables could be stored in a "common" format. JSON or whatever. That both the justfile and sysgo can consume. This is good enough though.

ajsutton replied:
Yeah I considered that but having the justfile be completely standalone is quite nice and makes the caching work easily.

Inphi reviewed (approved):
Looks good. Do consider whether we can avoid introduce the shim. Only setting up a kona trace config for tests that need it.

ajsutton committed:
Use kona v1.2.2

ajsutton committed:
Fix kona prestate URL

ajsutton committed:
Only set cannon-kona config for tests that opt-in.
Removes the need for run-kona-host.sh
